cmd/bnscli/cmd_key.go

```diff
@@ -55,6 +67,28 @@ created. This command fails if the private key file already exists.
 	return nil
 }
 
+// keygen returns a private key generated using given mnemonic and derivation
+// path.
+func keygen(mnemonic, derivationPath string) (ed25519.PrivateKey, error) {
+	if !bip39.IsMnemonicValid(string(mnemonic)) {
+		return nil, errors.New("invalid mnemonic")
```

Comment:
Can we test this validation for

* valid mnemonic 12 words
* valid mnemonic 15 words
* valid mnemonic 18 words
* valid mnemonic 21 words
* valid mnemonic 24 words
* invalid: mnemonic with wrong checksum (change last word to some other workd from the workdlist)
* invalid: additional whitespace around valid mnemonnic
* invalid: mnenomic that is valid in a language other than English

Reply:
The code processing the mnemonic is not ours and usually I would argue that we should not test it. But here it is https://github.com/iov-one/weave/pull/923/commits/4b19844a839fe8f091a0cb49952c968dd57a2342 

> invalid: mnemonic with wrong checksum (change last word to some other workd from the workdlist)
I do not understand this case, so I did not include it in the tests.

Reply:
> The code processing the mnemonic is not ours and usually I would argue that we should not test it.

We don't need to test every single code branch of that external library but we should at least cover our integration of it as well as the bahaviour we expect from the API.

> I do not understand this case, so I did not include it in the tests.

I'm glad we talk about it: A BIP39 mnemonic is not just a list of 12, 15, 18, 21 or 24 words from a 2048 element wordlist. It is encoded data plus a checksum. In a 12 word mnemonic for example, 11.64 words are arbitrary data and 0.36 words are a checksum.

e.g.

```
valid:
crack turtle seminar height entire subway motion rail pass seat violin scene

invalid (first word changed):
turtle turtle seminar height entire subway motion rail pass seat violin scene

invalid (last word changed):
crack turtle seminar height entire subway motion rail pass seat violin violin
```

Reply:
It seems as the implementation is not complete. 
https://github.com/tyler-smith/go-bip39/blob/master/bip39.go#L275-L294

1. It ignores whitespace by using `string.Fields` to extract words, which seems fair
2. It ensures that the number of words is in the right amount
3. It ensures that all words belong to the dictionary
4. It **does not** check the checksum

I think the best is to create a pull request with a fix.

Reply:
> I think the best is to create a pull request with a fix.

I agree. Good that we checked that ;)

Regarding 3.: the behaviour is fair indeed. We should just be explicit about it since not every implementations wirks that way. In this case we should have a test that verifies that mnemonic and menonic+whitespace leads to the same keypair.